Add "shibaken" generator for IMDS public keys in user-data

jpculp opened the pull request:
**Issue number:**
N/A


**Description of changes:**
This lets a new user-data block be generated with public keys available
in IMDS in the event that user-data has not been populated by the user.


**Testing done:**



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

hencrice committed:
Switched from TermLogger to SimpleLogger

Resolves #1157.

zmrow committed:
Add a vmware-dev variant

This change adds the necessary files for a VMWare development variant.
The files are currently a copy of the aws-dev variant but will diverge
over time.

Commit:
containerd: update to 1.4.3

Drop all backported patches and security fixes, since they are in the
new release.

Drop the SELinux patches, to take advantage of the improved support.

Remove the socat dependency, since it is no longer used for CRI port
forwarding.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
selinux-policy: modify default label for containers

This changes the default label from `container_t`, which is meant as
the unprivileged type, to `control_t`, which has higher privileges,
including the ability to write to the host API socket.

This is better aligned with the SELinux logic in containerd's CRI
plugin and in the docker daemon, where the label from lxc_contexts is
used for unprivileged containers, while privileged containers are not
explicitly labeled and follow the default transition rules.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
kubernetes: move kubelet args to a separate template

By using a systemd drop-in file for the kubelet args, we can control
them using a template, while leaving the main service unit in place
for dependency resolution.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
kubernetes-1.19: add network proxy support

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
kubernetes: add support for standalone mode

This also makes the API server, cluster name, and cluster certificate
settings optional, since the kubelet does not need these values when
running in standalone mode.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
kubernetes: add support for TLS auth

This enables the kubelet to use TLS for authentication instead of AWS
role credentials. If a bootstrap token is provided, the kubelet will
use it to perform TLS bootstrapping.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

etungsten committed:
settings, static-pods: new settings and helper for k8s static pods

Adds a set of new settings for managing k8s static pods.
Adds a new binary helper to manage k8s static pod manifests.

etungsten committed:
aws-k8s-*: specify 'staticPodPath' in kubelet config

Specify pod manifest path in kubelet config for all kubernetes variants.

zmrow committed:
host-containers: Don't fail if host containers aren't defined

If a user doesn't define any host containers, it only means that this
program has nothing to do.  This change prevents the program from
failing, which causes other boot services to fail.

Commit:
docs: update SELinux details in SECURITY_GUIDANCE.md

This documents the switch to running privileged containers with the
`control_t` label by default, and describes the implications.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

etungsten committed:
pluto: only build for kubernetes variants

We only need to build pluto for kubernetes variants

tjkirch committed:
Allow overriding the "pretty" name of the OS inside the image

webern committed:
metricdog: add https proxy environment variables

Use the proxy.env file to give HTTPS_PROXY and NO_PROXY values to
metricdog. metricdog's HTTP client automatically picks these up and uses
them.

webern committed:
metricdog: prepend scheme to https proxy value

reqwest does not support HTTPS_PROXY values that are not in URL format
(i.e. if there is no protocol scheme). On Bottlerocket we expect an
HTTPS_PROXY value with no scheme to be treated as http://.

Here we check the HTTPS_PROXY value, and if it has no scheme we reset
the value to prepend https://

webern committed:
updog: respect proxy settings

Adds https-proxy and no-proxy settings to updog's config file. These are
used to set HTTPS_PROXY and NO_PROXY environment variables for the
process at the beginning of the program which which ensures that they
are used by the underlying HTTP client.

webern committed:
makefile: add script runner bash to link-variant

Without this change, builds were failing on some systems.

webern committed:
Merge pull request #1323 from webern/makefile

makefile: add script runner bash to link-variant

zmrow commented on sources/api/shibaken/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+exclude = ["README.md"]
+
+[dependencies]
+apiclient = { path = "../apiclient" }
```
Take a quick spin through and make sure all these dependencies are actually needed

jpculp replied:
It looks like I need all of them except maybe http?

zmrow commented on sources/api/shibaken/README.md:
```diff
@@ -0,0 +1,17 @@
+## Introduction
+
+shibaken is called by sundog as a setting generator.
+Its sole parameter is the user-data provided by the api server.
```
The parameter is the host container user data setting.

jpculp replied:
No longer a parameter

zmrow commented on sources/api/shibaken/README.md:
```diff
@@ -0,0 +1,17 @@
+shibaken is called by sundog as a setting generator.
+Its sole parameter is the user-data provided by the api server.
+
+In the event that user-data is empty and has not been provided by the user, shibaken will generate
```
Can specify `...that host container user-data is empty...`.  You probably don't need to mention that it has not been provided since the setting being empty implies that. :)

zmrow commented on sources/api/shibaken/README.md:
```diff
@@ -0,0 +1,17 @@
+Its sole parameter is the user-data provided by the api server.
+
+In the event that user-data is empty and has not been provided by the user, shibaken will generate
+its own user-data by fetching authorized ssh keys from the AWS instance metadata service (IMDS).
```
I think I might reword this to say:

`shibaken will fetch and populate the host container's user-data with authorized keys from ...`

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+}
+
+/// Print usage message.
+fn usage() -> ! {
```
Let's move the usage and parse_args, down below run()

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+/// Print usage message.
+fn usage() -> ! {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!("Usage: {} user-data", program_name);
```
I think this is incorrect - we want the user to supply the setting  to fetch, so the invocation would look like:

`shibaken settings.host-containers.<container name>.user-data`

Take a peek at schnauzer's invocation in `sources/models/shared-defaults/defaults.toml`

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+/// Parses args for the setting key name.
+fn parse_args(mut args: env::Args) -> String {
+    let arg = args.nth(1).unwrap_or_else(|| usage());
+    if arg == "--help" || arg == "-h" {
```
Same here - we want to receive the setting name to fetch.

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+
+    info!("shibaken started");
+
+    info!("Fetching user-data metadata");
```
Fetching host-container user-data

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+    info!("shibaken started");
+
+    info!("Fetching user-data metadata");
+    let metadata_user_data = get_metadata(&setting_name, "user-data").await?;
```
Let's re-name this variable to just `user_data`.  `metadata_user_data` is sort of confusing.

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+/// Returns the value of a metadata key for a given data key, erroring if the value is not a
+/// string or is empty.
+async fn get_metadata(key: &str, meta: &str) -> Result<String> {
+    let uri = &format!("{}{}?keys={}", API_METADATA_URI_BASE, meta, key);
```
I believe we should be querying a setting here - not metadata.  `user-data` is a setting.

zmrow committed:
Merge pull request #1292 from zmrow/vmware-dev-variant

Add a vmware-dev variant

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+        }
+    );
+
+    // Metadata responses are of the form `{"data_key": METADATA}` so we pull out the value.
+    let mut response_map: HashMap<String, serde_json::Value> =
+        serde_json::from_str(&response_body).context(error::DeserializeJson)?;
+    let response_val = response_map
+        .remove(key)
+        .context(error::MissingMetadata { meta, key })?;
+
+    // Ensure it's a non-empty string
+    let response_str = response_val
+        .as_str()
+        .with_context(|| error::MetadataWrongType {
+            meta,
+            expected: "string",
+            value: response_val.to_string(),
+        })?;
+    let metadata = if !response_str.is_empty() {
+        response_str.to_string();
+    } else {
+        String::new();
+    };
```
I don't think we'll need to do any of this - we should be able to take the string we get back from the API, and if it's not empty, attempt to deserialize it into the structs you defined.

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+
+    let output = if metadata_user_data.is_empty() {
+
+        info!("No public keys available in user-data. Checking IMDS...");
```
This should probably read something like:

`No user-data exists for this container.  Populating user-data with keys from IMDS..."`

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+    let client = Client::new();
+    // Use IMDSv2 for accessing instance metadata.
+    let uri = IMDS_SESSION_TOKEN_ENDPOINT;
+    let imds_session_token = client
```
Take a peek at `early-boot-config`'s handling of session tokens.  There, it is split into a function and the errors are a bit more descriptive.  

`sources/api/early-boot-config/src/provider/aws.rs`  < `fetch_imds_session_token()`

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+        .context(error::ImdsText { uri })?;
+
+    info!("Retrieving list of available public keys from IMDS");
+    let public_key_list =
```
Let's put a short comment here explaining what the "list" is.

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+        get_text_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URL, &imds_session_token)?;
+    debug!("available public keys '{}'", &public_key_list);
+
+    info!("Generating endpoint urls to retrieve text of available public keys");
```
This can probably be `debug!()`.  It's an implementation detail related to fetching public keys.

etungsten committed:
migrations: add `add-static-pods` migration

Adds a new migration for new k8s static-pods settings

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+/// Returns a list of endpoint urls for the public keys in IMDS. Since IMDS returns the list of
+/// available public keys as '0=my-public-key', we need to strip the index from the list and
+/// embed it into the endpoint url.
+fn get_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<Url>> {
```
At this point, we're not getting anything from IMDS.  We can probably name this function like `build_public_key_endpoints()`

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+/// embed it into the endpoint url.
+fn get_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<Url>> {
+    let mut public_key_endpoints = Vec::new();
+    for available_key in public_key_list.lines() {
```
Is the output from IMDS newline delimited?

jpculp replied:
I believe so, but I'm not able to test it as I can't find a way to store a second key in IMDS.

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+    for available_key in public_key_list.lines() {
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<i32>().is_ok() {
```
I'm still wondering if this is even possible.  It's not a bad check I suppose.

jpculp replied:
The only way I could think of is if maybe someone put an = in their key name? Either way, doesn't hurt to check.

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<i32>().is_ok() {
+            let public_key_endpoint = &format!("{}{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
```
`public_key_endpoint_str`

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+        base64::encode(&user_data_json);
+
+    } else {
+        metadata_user_data;
+    };
```
I don't think we want this `else` condition here.  If the user-data exists, we don't need to do anything.

tjkirch committed:
Merge pull request #1282 from hencrice/ylc/1157

Switched from TermLogger to SimpleLogger

Commit:
Merge pull request #1318 from bcressey/default-container-label

modify default label for containers

webern committed:
logdog: add eni logs for k8s

tjkirch committed:
models: only run build.rs once, for host

tjkirch committed:
models: rerun if variant/mod links change

We don't expect these links to change outside of this script, but if they do,
we need to rerun this build.rs and reset the links properly.

tjkirch committed:
models: fix TOCTOU in build.rs by safely swapping links into place

Commit:
remove socat package

CRI has its own implementation for port forwarding now, and there are
no other packages that depend on this.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

webern committed:
Merge pull request #1327 from webern/logs

logdog: add eni logs for k8s

etungsten committed:
Merge pull request #1317 from etungsten/static-pods

k8s static pods support

jpculp committed:
Add "shibaken" generator to populate user-data with public keys from IMDS

This populates the admin container's user-data setting with public keys
available from IMDS in the event that user-data has not been populated
by the user.

webern committed:
docker: update to 19.03.15

tjkirch committed:
Merge pull request #1329 from tjkirch/model-link

models: link current variant more carefully

tjkirch committed:
buildsys: use argument vecs instead of string manipulation

Some arguments have internal spaces, like the list of packages in --build-arg
PACKAGES=a b c.  This was handled using a "|" marker that was later replaced by
spaces.  For other uses of --build-arg, like PRETTY_NAME, we don't want to
arbitrarily restrict the characters that can be used, in case someone wants "|"
in their OS name.

This change stores arguments directly in vectors, rather than strings that we
split on space, so that we can pass through arbitrary characters.

webern committed:
update k8s to 1.17.17

webern committed:
update k8s to 1.18.16

webern committed:
update k8s to 1.19.8

Commit:
containerd: fix volume mount labels for CRI

This fixes an issue where volume mounts received the `cache_t` label,
which is not writable by most processes.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
runc: upgrade to 1.0.0-rc93

Set the version and commit strings for `runc -v` output.

Drop the /dev/mqueue patch, which was fixed upstream:
  https://github.com/opencontainers/runc/pull/2558

Signed-off-by: Ben Cressey <bcressey@amazon.com>

sanu11 committed:
glibc: update to 2.33

tjkirch committed:
Merge pull request #1330 from tjkirch/pretty-name

Allow overriding the "pretty" name of the OS inside the image

Commit:
Merge pull request #1333 from webern/docker

docker: update to 19.03.15

Commit:
add migrations for kubelet standalone mode and TLS auth

Signed-off-by: Ben Cressey <bcressey@amazon.com>

arnaldo2792 committed:
ca-certificates: update to 2021.01.19

arnaldo2792 committed:
cni-plugins: update to 0.9.1

arnaldo2792 committed:
dbus-broker: update to 27

arnaldo2792 committed:
e2fsprogrs: update to 1.46.1

arnaldo2792 committed:
findutils: update to 4.8.0

arnaldo2792 committed:
iptables: update to 1.8.7

Commit:
Merge pull request #1337 from bcressey/k8s-1.19-proxy

add network proxy support to k8s 1.19

arnaldo2792 committed:
iputils: update to 20210202

arnaldo2792 committed:
kernel: update to 5.4.91

arnaldo2792 committed:
kmod: update to 28

arnaldo2792 committed:
libaudit: update to 3.0.1

arnaldo2792 committed:
libcap: update to 2.48

arnaldo2792 committed:
libnftnl: update to 1.1.9

arnaldo2792 committed:
libpcap: update to 1.10.0

arnaldo2792 committed:
procps: update to 3.3.17

arnaldo2792 committed:
strace: update to 5.11

webern committed:
Merge pull request #1335 from webern/k8s-updates

k8s updates

webern committed:
Merge pull request #1322 from webern/metricdog-proxy

metricdog: respect proxy settings

webern committed:
Merge pull request #1324 from webern/updog-proxy

updog: respect proxy settings

Commit:
Merge pull request #1338 from bcressey/standalone-mode

add support for kubelet in standalone mode and TLS auth

Commit:
Merge pull request #1336 from bcressey/containerd-and-runc

update containerd to 1.4.3, runc to 1.0.0-rc93

arnaldo2792 committed:
Merge pull request #1340 from arnaldo2792/updates

Update packages

etungsten committed:
defaults: bump control container default version to v0.4.2

Set default control container version to v0.4.2

etungsten committed:
migrations: control-container v0.4.2 default version migration

Migrates default version of control container from v0.4.1 to v0.4.2

gregdek committed:
Clarify use case for update repos (#1339)

Clarify use case for update repos

sanu11 committed:
Merge pull request #1334 from sanu11/develop

glibc: update to 2.33

etungsten committed:
Merge pull request #1344 from etungsten/control-container-migrate-v0.4.1

defaults, migrations: control container version bump to v0.4.2

jpculp committed:
migration-helpers: `AddMetadataMigration` for setting metadata

Adds new migration helper `AddMetadataMigration` to remove setting
metadata (like setting-generators).

jpculp committed:
migrations: add `add-shibaken` migration

Adds a new migration for shibaken setting-generator
